File: services/price-service/internal/adapters/base/xlsx.go
package base

import (
	"regexp"

	"github.com/kosarica/price-service/internal/parsers/xlsx"
	"github.com/kosarica/price-service/internal/types"
)

// XlsxAdapterConfig contains configuration for XLSX-based chain adapters
type XlsxAdapterConfig struct {
	BaseAdapterConfig
	ColumnMapping            xlsx.XlsxColumnMapping
	AlternativeColumnMapping *xlsx.XlsxColumnMapping
	// HasHeader indicates whether the first data row is a header
	HasHeader bool
	// HeaderRowCount is the number of rows to skip before data starts
	HeaderRowCount int
	// DefaultStoreIdentifier is used if not found in spreadsheet
	DefaultStoreIdentifier string
}

// BaseXlsxAdapter provides common XLSX parsing logic
type BaseXlsxAdapter struct {
	*BaseChainAdapter
	xlsxParser               *xlsx.Parser
	columnMapping            xlsx.XlsxColumnMapping
	altMapping               *xlsx.XlsxColumnMapping
	hasHeader                bool
	headerRowCount           int
	defaultStoreIdentifier   string
}

// NewBaseXlsxAdapter creates a new base XLSX adapter
func NewBaseXlsxAdapter(cfg XlsxAdapterConfig) (*BaseXlsxAdapter, error) {
	// Set XLSX file extension pattern
	if cfg.FileExtensionPattern == nil {
		cfg.FileExtensionPattern = regexp.MustCompile(`\.(xlsx|xls|XLSX|XLS)$`)
	}

	// Create base adapter
	base, err := NewBaseChainAdapter(cfg.BaseAdapterConfig)
	if err != nil {
		return nil, err
	}

	// Create XLSX parser
	parserOptions := xlsx.XlsxParserOptions{
		ColumnMapping:          &cfg.ColumnMapping,
		HasHeader:              cfg.HasHeader,
		HeaderRowCount:         cfg.HeaderRowCount,
		DefaultStoreIdentifier: cfg.DefaultStoreIdentifier,
		SkipEmptyRows:          true,
	}

	xlsxParser := xlsx.NewParser(parserOptions)
	if cfg.AlternativeColumnMapping != nil {
		xlsxParser.SetAlternativeMapping(cfg.AlternativeColumnMapping)
	}

	return &BaseXlsxAdapter{
		BaseChainAdapter:       base,
		xlsxParser:             xlsxParser,
		columnMapping:          cfg.ColumnMapping,
		altMapping:             cfg.AlternativeColumnMapping,
		hasHeader:              cfg.HasHeader,
		headerRowCount:         cfg.HeaderRowCount,
		defaultStoreIdentifier: cfg.DefaultStoreIdentifier,
	}, nil
}

// Parse parses XLSX content into normalized rows
// Tries primary column mapping first, then alternative if no valid rows
func (a *BaseXlsxAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	// Preprocess content (can be overridden by subclasses)
	processedContent := a.preprocessContent(content)

	// Get store identifier - prefer default if set (e.g., for national pricing)
	// Only fall back to filename extraction if no default is configured
	storeIdentifier := a.defaultStoreIdentifier
	if storeIdentifier == "" {
		storeIdentifier = a.extractStoreIdentifierFromFilename(filename)
	}

	// Parse with store identifier
	result, err := a.xlsxParser.ParseWithStoreID(processedContent, storeIdentifier)
	if err != nil {
		return nil, err
	}

	// Post-process results
	return a.postprocessResult(result), nil
}

// preprocessContent preprocesses content before parsing
// Can be overridden by subclasses
func (a *BaseXlsxAdapter) preprocessContent(content []byte) []byte {
	return content
}

// postprocessResult post-processes parse result
// Can be overridden by subclasses
func (a *BaseXlsxAdapter) postprocessResult(result *types.ParseResult) *types.ParseResult {
	return result
}

// GetColumnMapping returns the primary column mapping
func (a *BaseXlsxAdapter) GetColumnMapping() xlsx.XlsxColumnMapping {
	return a.columnMapping
}

// GetAlternativeMapping returns the alternative column mapping
func (a *BaseXlsxAdapter) GetAlternativeMapping() *xlsx.XlsxColumnMapping {
	return a.altMapping
}

// SetParserOptions updates the parser options
func (a *BaseXlsxAdapter) SetParserOptions(options xlsx.XlsxParserOptions) {
	a.xlsxParser.SetOptions(options)
}

// extractStoreIdentifierFromFilename extracts store identifier string from filename
func (a *BaseXlsxAdapter) extractStoreIdentifierFromFilename(filename string) string {
	baseName := a.fileExtensionPattern.ReplaceAllString(filename, "")

	cleanName := baseName
	for _, pattern := range a.filenamePrefixPatterns {
		cleanName = pattern.ReplaceAllString(cleanName, "")
	}

	cleanName = trimWhitespace(cleanName)

	if cleanName == "" {
		return a.fileExtensionPattern.ReplaceAllString(filename, "")
	}

	return cleanName
}

// trimWhitespace trims whitespace from string
func trimWhitespace(s string) string {
	return regexp.MustCompile(`^\s+|\s+$`).ReplaceAllString(s, "")
}

File: src/ingestion/adapters/base/xlsx.ts
import { ResultAsync } from "neverthrow";
import { fetchError, type FetchError } from "@/lib/errors";
import {
	newHeaderIndex,
	newNumericIndex,
	type XlsxColumnMapping,
	XlsxParser,
	type XlsxParserOptions,
} from "../../parsers/xlsx";
import type { ParseOptions, ParseResult } from "../../types";
import { type BaseAdapterConfig, BaseChainAdapter } from "./chain";

export interface XlsxAdapterConfig {
	baseConfig: BaseAdapterConfig;
	columnMapping: XlsxColumnMapping;
	alternativeColumnMapping?: XlsxColumnMapping;
	hasHeader: boolean;
	headerRowCount: number;
	defaultStoreIdentifier?: string;
}

export class BaseXlsxAdapter extends BaseChainAdapter {
	protected xlsxParser: XlsxParser;
	protected columnMapping: XlsxColumnMapping;
	protected altMapping?: XlsxColumnMapping;
	protected hasHeader: boolean;
	protected headerRowCount: number;
	protected defaultStoreIdentifier: string;

	constructor(cfg: XlsxAdapterConfig) {
		super({
			...cfg.baseConfig,
			fileExtensionPattern:
				cfg.baseConfig.fileExtensionPattern ?? /\.(xlsx|xls|XLSX|XLS)$/,
		});

		this.columnMapping = cfg.columnMapping;
		this.altMapping = cfg.alternativeColumnMapping;
		this.hasHeader = cfg.hasHeader;
		this.headerRowCount = cfg.headerRowCount;
		this.defaultStoreIdentifier = cfg.defaultStoreIdentifier ?? "";

		this.xlsxParser = new XlsxParser({
			columnMapping: cfg.columnMapping,
			hasHeader: cfg.hasHeader,
			headerRowCount: cfg.headerRowCount,
			defaultStoreIdentifier: cfg.defaultStoreIdentifier,
			skipEmptyRows: true,
		});
		this.xlsxParser.setAlternativeMapping(cfg.alternativeColumnMapping);
	}

	parse(
		content: Buffer,
		filename: string,
		_options?: ParseOptions,
	): ResultAsync<ParseResult, FetchError> {
		return ResultAsync.fromPromise(
			Promise.resolve().then(() => {
				const processed = this.preprocessContent(content);
				const storeIdentifier =
					this.defaultStoreIdentifier ||
					this.extractStoreIdentifierFromFilename(filename);
				const result = this.xlsxParser.parseWithStoreId(
					processed,
					storeIdentifier,
				);
				return result;
			}),
			(e) =>
				fetchError({
					url: filename,
					message: e instanceof Error ? e.message : "XLSX parse failed",
					retryable: false,
					attempts: 0,
					cause: e,
				}),
		)
			.map((result) => this.postprocessResult(result));
	}

	protected preprocessContent(content: Buffer): Buffer {
		return content;
	}

	protected postprocessResult(result: ParseResult): ParseResult {
		return result;
	}

	setParserOptions(options: XlsxParserOptions): void {
		this.xlsxParser.setOptions(options);
	}

	protected extractStoreIdentifierFromFilename(filename: string): string {
		const baseName = filename.replace(this.fileExtensionPattern, "");
		let cleanName = baseName;
		for (const pattern of this.filenamePrefixPatterns) {
			cleanName = cleanName.replace(pattern, "");
		}
		cleanName = cleanName.trim();
		if (!cleanName) {
			return filename.replace(this.fileExtensionPattern, "");
		}
		return cleanName;
	}
}

export { newHeaderIndex, newNumericIndex };
